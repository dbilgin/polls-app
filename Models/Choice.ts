export class Choice {
  choice: string;
  url: string;
  votes: number;

  constructor(choice: string, url: string, votes: number) {
    this.choice = choice;
    this.url = url;
    this.votes = votes;
  }
}
