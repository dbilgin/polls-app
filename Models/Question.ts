import {Choice} from './Choice';

export interface Question {
  id?: number;
  question: string;
  published_at: Date;
  url: string;
  choices: Choice[];
}
