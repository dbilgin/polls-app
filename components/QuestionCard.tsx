import React from 'react';
import {StyleSheet, View, Text, Image, ImageSourcePropType} from 'react-native';
import {Question} from '../Models/Question';

interface CardProps {
  q: Question;
  image: ImageSourcePropType;
}

const QuestionCard: React.FC<CardProps> = ({q, image}) => {
  function sum(obj: number[]) {
    let finalSum = 0;
    for (const el of obj) {
      finalSum += el;
    }
    return finalSum;
  }

  return (
    <View style={styles.cardContainer}>
      <Image
        style={{flex: 0.5, width: '100%'}}
        resizeMode={'contain'}
        source={image}
      />
      <View style={{padding: 24, flex: 1}}>
        <Text style={styles.title}>{q.question}</Text>
        <Text style={styles.subtitle}>{q.url}</Text>
        <Text style={styles.dateText}>
          Published at: {new Date(q.published_at).toLocaleDateString()}
        </Text>
        <Text style={styles.bottomText}>{`This question has ${
          q.choices.length
        } choices and a total of ${sum(
          q.choices.map((x) => x.votes),
        )} votes. Press here now to vote!`}</Text>
      </View>
    </View>
  );
};

export default React.memo(QuestionCard);

const styles = StyleSheet.create({
  bottomText: {
    fontSize: 24,
    color: 'black',
    marginTop: 24,
  },
  dateText: {
    fontSize: 16,
    color: '#505b7d',
    marginTop: 12,
  },
  subtitle: {
    fontSize: 12,
    color: '#505b7d',
    marginTop: 12,
  },
  title: {
    fontSize: 24,
    color: '#505b7d',
  },
  cardContainer: {
    height: 500,
    width: '80%',
    margin: 20,
    borderWidth: 0,
    borderRadius: 18,
    justifyContent: 'center',
    alignSelf: 'center',
    backgroundColor: 'white',

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,

    elevation: 8,
  },
});
