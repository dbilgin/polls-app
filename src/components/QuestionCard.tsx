import React from 'react';
import {
  StyleSheet,
  View,
  Text,
  Image,
  ImageSourcePropType,
  TouchableOpacity,
  SafeAreaView,
  Alert,
} from 'react-native';
import {Question} from '../../Models/Question';
import {sum} from '../utils';
import {RootStackParamList} from '../../App';
import {StackNavigationProp} from '@react-navigation/stack';

interface CardProps {
  q: Question;
  image: ImageSourcePropType;
  navigation: StackNavigationProp<RootStackParamList, 'QuestionListScreen'>;
}

const QuestionCard: React.FC<CardProps> = ({q, image, navigation}) => {
  const navigateToDetail = () => {
    if (q.id) {
      navigation.navigate('QuestionDetailScreen', {questionId: q.id});
    } else {
      Alert.alert('An error has occurred.');
    }
  };

  return (
    <SafeAreaView style={{flex: 1}}>
      <TouchableOpacity onPress={navigateToDetail} style={styles.cardContainer}>
        <Image
          style={{flex: 0.5, width: '100%'}}
          resizeMode={'contain'}
          source={image}
        />
        <View style={{padding: 24, flex: 1}}>
          <Text style={styles.title}>{q.question}</Text>
          <Text style={styles.subtitle}>{q.url}</Text>
          <Text style={styles.dateText}>
            Published at: {new Date(q.published_at).toLocaleDateString()}
          </Text>
          <Text style={styles.bottomText}>{`This question has ${
            q.choices.length
          } choices and a total of ${sum(
            q.choices.map((x) => x.votes),
          )} votes. Press here now to vote!`}</Text>
        </View>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default QuestionCard;

const styles = StyleSheet.create({
  bottomText: {
    fontSize: 24,
    color: 'black',
    marginTop: 24,
  },
  dateText: {
    fontSize: 16,
    color: '#505b7d',
    marginTop: 12,
  },
  subtitle: {
    fontSize: 12,
    color: '#505b7d',
    marginTop: 12,
  },
  title: {
    fontSize: 24,
    color: '#505b7d',
  },
  cardContainer: {
    height: 500,
    width: '80%',
    margin: 20,
    borderWidth: 0,
    borderRadius: 18,
    justifyContent: 'center',
    alignSelf: 'center',
    backgroundColor: 'white',

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,

    elevation: 8,
  },
});
