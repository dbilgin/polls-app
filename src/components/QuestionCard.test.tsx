import React from 'react';
import {shallow} from 'enzyme';
import QuestionCard from './QuestionCard';
import {Question} from '../../Models/Question';
import {SafeAreaView, TouchableOpacity, View, Image, Text} from 'react-native';

const createTestProps = () => ({
  navigation: {
    navigate: jest.fn(),
  },
  image: require('../images/1.png'),
});

describe('QuestionCard', () => {
  describe('Rendering', () => {
    let props: any;
    beforeEach(() => {
      props = createTestProps();
    });

    it('should match to snapshot', () => {
      const question: Question = {
        id: 1,
        question: 'question',
        choices: [],
        published_at: new Date(),
        url: '/url',
      };

      const component = shallow(<QuestionCard q={question} {...props} />);
      expect(component).toMatchSnapshot();
    });
  });

  describe('UI Placements', () => {
    let props: any;
    beforeEach(() => {
      props = createTestProps();
    });

    it('should have correct count of components', () => {
      const question: Question = {
        id: 1,
        question: 'question',
        choices: [],
        published_at: new Date(),
        url: '/url',
      };

      const component = shallow(<QuestionCard q={question} {...props} />);

      expect(component.find(SafeAreaView).length).toBe(1);
      expect(component.find(TouchableOpacity).length).toBe(1);
      expect(component.find(Image).length).toBe(1);
      expect(component.find(View).length).toBe(1);
      expect(component.find(Text).length).toBe(4);
    });
  });

  describe('Interaction', () => {
    let props: any;
    beforeEach(() => {
      props = createTestProps();
    });

    it('should navigate', async () => {
      const question: Question = {
        id: 1,
        question: 'question',
        choices: [],
        published_at: new Date(),
        url: '/url',
      };

      const component = shallow(<QuestionCard q={question} {...props} />);

      // @ts-ignore
      await component.find(TouchableOpacity).first().props().onPress();
      expect(props.navigation.navigate).toHaveBeenCalled();
    });

    it('should not navigate', async () => {
      const question: Question = {
        question: 'question',
        choices: [],
        published_at: new Date(),
        url: '/url',
      };

      const component = shallow(<QuestionCard q={question} {...props} />);

      // @ts-ignore
      await component.find(TouchableOpacity).first().props().onPress();
      expect(props.navigation.navigate).not.toHaveBeenCalled();
    });
  });
});
