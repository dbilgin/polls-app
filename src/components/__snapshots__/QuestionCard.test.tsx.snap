// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionCard Rendering should match to snapshot 1`] = `
<SafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "white",
        "borderRadius": 18,
        "borderWidth": 0,
        "elevation": 8,
        "height": 500,
        "justifyContent": "center",
        "margin": 20,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 4.65,
        "width": "80%",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../src/images/1.png",
        }
      }
      style={
        Object {
          "flex": 0.5,
          "width": "100%",
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "padding": 24,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 24,
          }
        }
      >
        question
      </Text>
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 12,
            "marginTop": 12,
          }
        }
      >
        /url
      </Text>
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 16,
            "marginTop": 12,
          }
        }
      >
        Published at: 
        6/7/2020
      </Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 24,
            "marginTop": 24,
          }
        }
      >
        This question has 0 choices and a total of 0 votes. Press here now to vote!
      </Text>
    </View>
  </TouchableOpacity>
</SafeAreaView>
`;
